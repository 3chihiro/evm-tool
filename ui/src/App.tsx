import React, { useCallback, useMemo, useState, useEffect } from 'react'
import GanttCanvas from './components/GanttCanvas'
import TaskTable from './components/TaskTable'
import EvmCard from './components/EvmCard'
import type { TaskRow } from '../../evm-mvp-sprint1/src.types'
import { parseCsvTextBrowser } from '../../src/adapters'
import { createUseHistory, type Command } from './lib/history'

const useTasksHistory = createUseHistory<TaskRow[]>([])

export default function App() {
  const hist = useTasksHistory()
  const tasks = hist.present
  const [errors, setErrors] = useState<string[]>([])
  const [selectedIds, setSelectedIds] = useState<(string | number)[]>([])

  const onFileChange = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
    const f = e.target.files?.[0]
    if (!f) return
    const reader = new FileReader()
    reader.onload = () => {
      const text = String(reader.result ?? '')
      const res = parseCsvTextBrowser(text)
      hist.set(res.tasks)
      setErrors(res.errors.map((er) => `Row ${er.row} ${er.column ?? ''} ${er.message}`))
      setSelectedIds([])
    }
    reader.readAsText(f, 'utf-8')
  }, [])

  const projectName = useMemo(() => tasks[0]?.projectName ?? '', [tasks])

  const onExportPDF = useCallback(async () => {
    if (!('appBridge' in window) || typeof window.appBridge.exportPDF !== 'function') {
      alert('Electron環境でのみPDF出力が利用できます。')
      return
    }
    const saved = await window.appBridge.exportPDF({ landscape: false })
    if (saved) alert(`PDFを保存しました: ${saved}`)
  }, [])

  return (
    <div className="app-grid">
      <header className="header">
        <h1>EVM Tool UI Shell</h1>
        <div style={{ marginLeft: '16px', fontSize: 12, color: '#666' }}>{projectName && `（${projectName}）`}</div>
        <div style={{ marginLeft: 'auto', display: 'flex', gap: 8, alignItems: 'center' }}>
          <input type="file" accept=".csv" onChange={onFileChange} />
          <button onClick={onExportPDF}>PDF出力</button>
        </div>
      </header>
      <section className="gantt">
        <GanttCanvas
          tasks={tasks}
          onTasksChange={(cmd: Command<TaskRow[]>) => hist.run(cmd)}
          selectedIds={selectedIds}
          onSelect={setSelectedIds}
        />
      </section>
      <section className="tasks">
        <TaskTable tasks={tasks} />
      </section>
      <aside className="evm">
        <EvmCard tasks={tasks} />
        {errors.length > 0 && (
          <div style={{ marginTop: 8 }}>
            <div className="panel-title">インポートエラー</div>
            <ul style={{ margin: 0, paddingLeft: 16 }}>
              {errors.map((e, i) => (
                <li key={i} style={{ color: '#b00', fontSize: 12 }}>{e}</li>
              ))}
            </ul>
          </div>
        )}
      </aside>
      <footer className="status">
        <div>
          選択中 {selectedIds.length} 件 ・ Escでキャンセル
        </div>
        <div>
          <button className="btn" disabled={selectedIds.length === 0} onClick={() => setSelectedIds([])}>選択解除</button>
        </div>
      </footer>
    </div>
  )
}
