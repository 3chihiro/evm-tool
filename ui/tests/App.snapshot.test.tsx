import { describe, it, expect } from 'vitest'
import { render } from '@testing-library/react'
import React from 'react'
import App from '../src/App'

describe('App snapshot', () => {
  it('renders three panes', () => {
    const { asFragment, getByText } = render(<App />)
    expect(getByText('EVM Tool UI Shell')).toBeTruthy()
    expect(getByText('ガント（ダミー）')).toBeTruthy()
    expect(getByText('タスク一覧')).toBeTruthy()
    expect(getByText('EVM（ダミー）')).toBeTruthy()
    expect(asFragment()).toMatchSnapshot()
  })
})

