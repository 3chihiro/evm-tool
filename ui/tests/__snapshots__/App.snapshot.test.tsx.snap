// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App snapshot > renders three panes 1`] = `
<DocumentFragment>
  <div
    class="app-grid"
  >
    <header
      class="header"
    >
      <h1>
        EVM Tool UI Shell
      </h1>
    </header>
    <section
      class="gantt"
    >
      <div>
        <div
          class="panel-title"
        >
          ガント（ダミー）
        </div>
        <canvas
          style="width: 100%; height: 240px;"
        />
      </div>
    </section>
    <section
      class="tasks"
    >
      <div>
        <div
          class="panel-title"
        >
          タスク一覧
        </div>
        <table
          class="table"
        >
          <thead>
            <tr>
              <th>
                名前
              </th>
              <th>
                開始
              </th>
              <th>
                終了
              </th>
              <th>
                %
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                企画
              </td>
              <td>
                2024-01-01
              </td>
              <td>
                2024-01-05
              </td>
              <td
                style="text-align: right;"
              >
                60
              </td>
            </tr>
            <tr>
              <td>
                設計
              </td>
              <td>
                2024-01-03
              </td>
              <td>
                2024-01-10
              </td>
              <td
                style="text-align: right;"
              >
                30
              </td>
            </tr>
            <tr>
              <td>
                実装
              </td>
              <td>
                2024-01-08
              </td>
              <td>
                2024-01-15
              </td>
              <td
                style="text-align: right;"
              >
                10
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </section>
    <aside
      class="evm"
    >
      <div
        class="card"
      >
        <div
          class="panel-title"
        >
          EVM（ダミー）
        </div>
        <div
          class="evm-grid"
        >
          <div
            class="evm-item"
          >
            <div
              class="evm-label"
            >
              PV
            </div>
            <div
              class="evm-value"
            >
              ￥1,000,000
            </div>
          </div>
          <div
            class="evm-item"
          >
            <div
              class="evm-label"
            >
              EV
            </div>
            <div
              class="evm-value"
            >
              ￥800,000
            </div>
          </div>
          <div
            class="evm-item"
          >
            <div
              class="evm-label"
            >
              AC
            </div>
            <div
              class="evm-value"
            >
              ￥900,000
            </div>
          </div>
          <div
            class="evm-item"
          >
            <div
              class="evm-label"
            >
              SV
            </div>
            <div
              class="evm-value"
            >
              -￥200,000
            </div>
          </div>
          <div
            class="evm-item"
          >
            <div
              class="evm-label"
            >
              CV
            </div>
            <div
              class="evm-value"
            >
              -￥100,000
            </div>
          </div>
          <div
            class="evm-item"
          >
            <div
              class="evm-label"
            >
              SPI
            </div>
            <div
              class="evm-value"
            >
              0.80
            </div>
          </div>
          <div
            class="evm-item"
          >
            <div
              class="evm-label"
            >
              CPI
            </div>
            <div
              class="evm-value"
            >
              0.89
            </div>
          </div>
        </div>
      </div>
    </aside>
  </div>
</DocumentFragment>
`;
